### DISCUSSION TITLE
Suggested Approach: Multiple Friction values on a single Surface

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#73] [Q&A] [answered]
With the implementation of the vehicle controller, I was looking into how to shift friction between a vehicle on/off the road.

I implemented the height-field shape which lets me index a different PhysicsMaterial per pixel to wire this up, but solving the initial task seems to require some form of using callbacks and SubShape collision without an easy way to change friction.

The main two methods I've see are:

1) Wire up 2 different bodies with appropriate frictions and just rely on the existing automatic behavior
2) Somehow test each update and modify the wheels friction itself based on their current in-contact models. Possibly the slip-friction arrays?

My implementation ideas so far have been either rendering as polygon meshes or heightfields, and then apply the race-track itself as an epsilon elevated version of the map (ex: 0.0001 y-displaced) to 'paint on' the track over the surface and ensure the right friction material is hit first by the vehicle. An alternative to the elevated map is to use the no-collision setting and create actual holes in both the maps so they stitch together exactly to a single surface.

What is the ideal way to have a terrain with different frictions, or possibly a way to take a single surface/plane and change friction based on location?

**Comment by LeXXik** [comment]
In our racing games, once the level design is complete, the track is simply exported as a separate mesh. Then grass and asphalt may have different properties. I'm all in for the features, but this sounds like a solution that is looking for a problem. I just can't recall a game out there that is using such an approach.

**Reply by PhoenixIllusion** [reply]
This was not intended as a feature proposal, but requesting suggestions on the best approach. Your description sounds similar to one of my options I mentioned, where every time a surface friction value changes, you create a different physics body mesh.

My understanding on the documentation is that there is a way to wire up physics materials so that different parts of a polygon structure can have different collision indicators, but that if those are wired as a single body, they can only have 1 friction.

Ex:
If a theoretical Quad surface has 4 different frictions or 10 different frictions over a gradient, is the suggested best approach for it to be to break it into 4 or 10 independent collision bodies with independent frictions, or is there a better approach?

If you had a complex mesh, is the only way to change friction along material surfaces in it to apply a filter in your code to break the triangle soup into X different buckets, 1 per friction value, and then generate X different collision bodies?

I have updated the title to state this is suggested approaches

**Comment by jrouwe** [comment]
Hello,

For meshes vs dynamic bodies, there is a demo in C++:

https://github.com/jrouwe/JoltPhysics/blob/8691a4ff82133b2770e31ee6ce73507a9db953bb/Samples/Tests/General/FrictionPerTriangleTest.cpp#L56-L95

The required functionality for this is exposed in JS as well (instead of ```MyMaterial``` you would use a regular ```PhysicsMaterial``` and e.g. give it a property ```friction``` in JS).

For meshes vs vehicles there is another callback:

https://github.com/jrouwe/JoltPhysics/blob/8691a4ff82133b2770e31ee6ce73507a9db953bb/Jolt/Physics/Vehicle/VehicleConstraint.h#L89-L94

This could be used in a similar way as the first example. The required function is not yet exposed to JS though and it would be the first time that a ```std::function``` is exposed so I haven't yet figured out how that works with the WebIDL bindings.

**Reply by PhoenixIllusion** [reply]
I think a cleanest immediate way I can think to do that is a JS callback class like we do for ContactListeners. Looking into it, WebIDL vs EmBind both have different ease of use of std::function and I'm not sure the cleanest way to support everything using WASM function pointers, although that is most flexible.

As lowest LoE and without too much overhead, I think we can use the `using` patterns to fully type the entire std::function into a single interface-name & WebIDL type. 

The class would have the expected friction method defined and exposed in JS for the friction combination and float return, and the class would also have a method that returns the fully WebIDL using-Typed std::function to pass to the Vehicle Constraint attribute

Pseudo-code: 
interface VehicleCombineFrictionEM {
  float CombineFriction(... ...);
  StdFunctionCombineFriction GetFunction();
};

**Reply by jrouwe** [reply]
Ok, I'll type something up

**Comment by jrouwe** [comment, marked as answer]
This is now released as part of 0.11.0, see example:

https://github.com/jrouwe/JoltPhysics.js/blob/75be070fae1bb8c57b67403659893aeb92f0f36b/Examples/vehicle_wheeled.html#L211-L220

**Reply by PhoenixIllusion** [reply]
That looks nice. The PreStepCallback looks like a good place to apply the Vehicle Input in of my BabylonJS plugin since the engine doesn't otherwise let my plugin easily hook into the actual physics step cycle with guaranteed delta times without a bit of a hassle.

Technically this new callback should allow for one of my initial ideas of bypassing the physics material setup and using a float data-texture lookup. During the friction function, grab the wheel and collided body and calculate a UV offset into a 2D array to pull a dynamic friction value for that surface without needing to generate frictions inside the physic system itself.

Maybe even making myself an actual friction-texture, allowing the actual track to be super low res instead of needing a high density mesh purely for non-collision physics attribute purposes (adding an oil slick or patch of ice)

**Reply by jrouwe** [reply]
> The PreStepCallback looks like a good place to apply the Vehicle Input in of my BabylonJS plugin since the engine doesn't otherwise let my plugin easily hook into the actual physics step cycle with guaranteed delta times without a bit of a hassle.

Yes, this should work.

> Technically this new callback should allow for one of my initial ideas of bypassing the physics material setup and using a float data-texture lookup.

I think this should be possible too. By the time you receive this callback, all collision checks have been performed so you can get whatever you need from the wheels and you don't need to use PhysicsMaterials if that's not convenient.
